orttraining/orttraining/test/python/checkpoint/_test_helpers.py

```diff
@@ -102,6 +104,93 @@ def create_orttrainer_and_load_checkpoint(device, trainer_opts, checkpoint_dir,
+
+    - Instantiates the ORTTrainer with given input trainer_opts configuration for a simple transformer model
+    - Runs train_step on the trainer so the trainer onnx graph is initialized
+    - Returns the trainer state_dict and the pytorch model
+    """
+    seed = 1
+    torch.manual_seed(seed)
+    set_seed(seed)
+
+    # PyTorch transformer model setup
+    learning_rate = 1e-10
+    optim_config = optim.LambConfig(lr=learning_rate) if use_lamb else optim.AdamConfig(lr=learning_rate)
+    model, model_desc, loss_fn, batcher_fn, train_data, _, _ = _load_pytorch_transformer_model(device)
+    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, loss_fn=loss_fn, options=orttrainer.ORTTrainerOptions(trainer_opts))
+
+    _train(trainer, train_data, batcher_fn)
+
+    model_state = trainer._training_session.get_model_state(include_mixed_precision_weights=False)
+    opt_state = trainer._training_session.get_optimizer_state()
+
+    return model_state, opt_state
+
+def create_initialized_orttrainer(device, trainer_opts, use_lamb=True):
+    seed = 1
+    torch.manual_seed(seed)
+    set_seed(seed)
+
+    learning_rate = 1e-10
+    optim_config = optim.LambConfig(lr=learning_rate) if use_lamb else optim.AdamConfig(lr=learning_rate)
+    model, model_desc, loss_fn, batcher_fn, train_data, _, _ = _load_pytorch_transformer_model(device)
+    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, loss_fn=loss_fn, options=orttrainer.ORTTrainerOptions(trainer_opts))
+
+    _train(trainer, train_data, batcher_fn)
+
+    return trainer
+
+def verify_model_state(trainer, expected_state_dict, is_mixedprecision):
+    actual_model_state = trainer._training_session.get_model_state(include_mixed_precision_weights=True)
+    
+    assert len(expected_state_dict['fp32_param']) == len(actual_model_state['full_precision']), \
+        "expected and actual should have same number of tensors"
+    for weight_name, tensor in expected_state_dict['fp32_param'].items():
+        if not weight_name in actual_model_state['full_precision']:
+            assert '_view_' in weight_name, \
+                "only zero shared weight may not match name"
+            shape = expected_state_dict['fp32_param'][weight_name].shape
+            weight_name = weight_name.split('_view_')[0]
+        assert_allclose(tensor, actual_model_state['full_precision'][weight_name])
+
+    if is_mixedprecision:
+        assert 'mixed_precision' in actual_model_state.keys(), "missing 'mixed_precision' key in mixed precision run"
+        assert len(expected_state_dict['fp16_param']) == len(actual_model_state['mixed_precision']), \
+            "expected and actual should have same number of tensors"
+        for weight_name, tensor in expected_state_dict['fp16_param'].items():
+            weight_name = weight_name.split('_fp16')[0]
+            assert_allclose(tensor, actual_model_state['mixed_precision'][weight_name], atol=global_fp16_fp32_atol)
```

Comment:
is this tolerance required even for comparing fp16 with fp16?

Reply:
Changed